src/plugins/data/public/search/types.ts

```diff
@@ -84,3 +98,30 @@ export interface SearchServiceStartDependencies {
   fieldFormats: AggsStartDependencies['fieldFormats'];
   indexPatterns: DataViewsContract;
 }
+
+/**
+ * Format of warnings of failed shards or internal ES timeouts that surface from search responses
+ * @public
+ */
+export interface SearchResponseWarning {
+  /**
+   * type:  for handling the warning in logic
+   */
+  type: 'timed_out' | 'generic_shard_warning' | estypes.ShardFailure['reason']['reason'];
+  /**
+   * isTimeout: true for general internal ES timeout warning
+   */
+  isTimeout?: boolean;
+  /**
+   * isTimeout: true for shard-specific internal ES warning
+   */
+  isShardFailure?: boolean;
+  /**
+   * message: failure reason from ES
+   */
+  message: string;
+  /**
+   * text: human-friendly error message
+   */
+  text?: string;
```

Comment:
i don;t think we need to separate between text and message

Reply:
When the warning's `isShardFailure = true`, a message is needed for the `ShardFailureOpenModalButton`

This is also why we need `isShardFailure`, otherwise the logic in src/plugins/data/public/search/fetch/handle_response.tsx will get convoluted (give the shard failure modal to any warning that isn't a timeout warning).

Reply:
the isShardFailure can be inferred from type. I understand now why we need the text however.